tests/publiccloud/sles4sap.pm

```diff
@@ -27,15 +37,187 @@ sub upload_ha_sap_logs {
+    foreach my $cmd (@cmds) {
+        # Each command execution has its own timeout, so we need to reset the counter
+        my $starttime = time;
+
+        # Check for cluster/resources status and exit loop when needed
+        my $ret;
+        while ($ret = $self->{my_instance}->run_ssh_command(cmd => "sudo $cmd", rc_only => 1)) {
+            # Otherwise wait a while if timeout is not reached
+            if (time - $starttime < $timeout) {
+                sleep 5;
+            }
+            else {
+                # Cluster doesn't start properly, show state before dying
+                $self->run_cmd(cmd => $crm_mon_cmd);
+                die "Cluster/resources did not start within $timeout seconds (cmd='$cmd')";
+            }
+        }
+
+        # $ret need to be defined to ensure a correct exit code
+        if (!defined $ret) {
+            # Cluster doesn't start properly, show state before dying
+            $self->run_cmd(cmd => $crm_mon_cmd);
+            die 'Cluster/resources check did not exit properly';
+        }
+    }
+}
+
+=head2 run_cmd
+    run_cmd(cmd => 'command', [timeout => 60]);
+
+Runs a command C<cmd> via ssh in the given VM and log the output.
+All commands are executed through C<sudo>.
+=cut
+sub run_cmd {
+    my ($self, %args) = @_;
+    die('Argument <cmd> missing') unless ($args{cmd});
+    my $timeout = bmwqemu::scale_timeout($args{timeout} // 60);
+    my $title   = $args{title} // $args{cmd};
+    $title =~ s/[[:blank:]].+// if !defined $args{title};
+
+    my $out = $self->{my_instance}->run_ssh_command(cmd => "sudo $args{cmd}", timeout => $timeout);
+    record_info("$title output - $self->{my_instance}->{instance_id}", $out);
+}
+
+=head2 fence_node
+
+    fence_node([hostname => $hostname], [timeout => 60]);
+
+Fence a HA node using C<systemctl poweroff>.
+This command terminates when the node is back.
+=cut
+sub fence_node {
+    my ($self, %args) = @_;
+    my $timeout = bmwqemu::scale_timeout($args{timeout} // 60);
+    $args{hostname} //= $self->{my_instance}->{instance_id};
+
+    record_info('Fencing info', 'Fencing done with systemctl poweroff');
+    $self->{my_instance}->run_ssh_command(cmd => "sudo systemctl poweroff --force", timeout => 0);
+
+    # skip the one minute waiting
+    sleep 60;
+    my $start_time = time();
+
+    # Wait till ssh disappear
+    while ((time() - $start_time) < $timeout) {
+        last unless (defined($self->{my_instance}->wait_for_ssh(timeout => 1, proceed_on_failure => 1)));
+    }
+    my $fencing_time = time() - $start_time;
+    die("Waiting for fencing failed!") unless ($fencing_time < $timeout);
+    record_info('Fencing finished!', "Fencing done in ${fencing_time}s");
+
+    # We have to wait "a little" to ensure that resources are moved to the other node
+    sleep $timeout;
+
+    # Wait for the node to restart
+    record_info("Restart $args{hostname}", 'Restart done using CSP API');
+    $self->{my_instance}->start(timeout => $timeout);
```

Comment:
I guess this is why we face problems with azure, plz take a look to https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/10115

Reply:
Exactly! I will test your suggestion, thx!